### DISCUSSION TITLE
where is the `on_server_ready`?

### DISCUSSION THREAD

**Discussion by nyngwang** [mason-org/mason.nvim#40] [Q&A] [answered]
Try to migrate from installer. My old codes are:

```lua
    require('nvim-lsp-installer').on_server_ready(
      function (server)
        local has_config, config = pcall(require, 'lsp_config.'..server.name)
        if not has_config then
          server:setup(require('lsp_config.default')) return end
        server:setup(config)
      end
    )
```

how should I do the same using mason?

**Comment by williamboman** [comment, marked as answer]
Hello! The `on_server_ready` function was [deprecated a while back](https://github.com/williamboman/nvim-lsp-installer/discussions/636), mostly due to how it wrapped lspconfig and was causing confusion. There is a similar alternative in the `mason-lspconfig` extension (to be documented), try this:

```lua
require("mason").setup()
require("mason-lspconfig").setup()

-- the above is enough, but if you want to replicate the "on_server_ready" behaviour
-- where your installed servers are setup "automatically" you can do the following
require("mason-lspconfig").setup_handlers {
    -- default handler - setup with default settings
    function (server_name)
        lspconfig[server_name].setup {}
    end,
   -- you can override the default handler by providing custom handlers per server
   ["jdtls"] = function ()
       // do something with the nvim-jdtls plugin instead
   end
}
```

Note that the `require("mason-lspconfig").setup_handlers` functions is not stable yet

**Reply by ghost** [reply]
Hello, sorry to be off-topic, but I didn't want to create another discussion for this. I haven't used neovim in a long time and I still have a somewhat clusterfuck of a configuration based around nvim-lsp-installer's on_server_ready(). At this point, should I just wait for mason.nvim to mature a bit and switch directly to it instead of getting rid of on_server_ready() now ?

**Reply by williamboman** [reply]
That's up to you. I'd recommend trying out mason.nvim now, I don't expect most of these things to change much (or at all, really), but I don't want to make any promises just yet, need to let it settle for a while.

**Reply by ghost** [reply]
I'm unsure how to do some stuff (I never touched lua outside configuring nvim) for example would something like this be possible with mason.nvim currently ? 
My ideal case would be to eventually have all the external stuff (lsp / linter / formatter) installed / uninstalled automatically based on lists and then have their commands or some other properties passed where needed (for example here to null-ls).

```lua
  local vim = vim
  local null_ls = require("null-ls")
  local format_install = require("format-installer")
  local util = require("util")

  local formatters = { "shfmt", "stylua" }

  for _, name in ipairs(formatters) do
      local formatter_is_found = format_install.exists(name)
      if formatter_is_found then
          if not format_install.is_installed(name) then
              format_install.install_formatter(name)
          end
      end
  end -- automatically install all formatters

  for _, formatter in ipairs(format_install.get_installed_formatters()) do
      if not util.set_contains_val(formatters, formatter.name) then
          format_install.uninstall_formatter(formatter.name)
      end
  end -- automatically uninstall all not needed formatters

local initial_sources = {}
  for _, formatter in ipairs(format_install.get_installed_formatters()) do
      local config = { command = formatter.cmd }
      table.insert(initial_sources, null_ls.builtins.formatting[formatter.name].with(config))
  end

  table.insert(initial_sources, null_ls.builtins.code_actions.shellcheck) -- shell
  table.insert(initial_sources, null_ls.builtins.formatting.shellharden) --shell
  table.insert(initial_sources, null_ls.builtins.diagnostics.selene) -- lua

  null_ls.setup({
      sources = initial_sources,
      on_attach = function(client)
          if client.resolved_capabilities.document_formatting then
              vim.cmd([[
              augroup LspFormatting
                  autocmd! * <buffer>
                  autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting_sync()
              augroup END
              ]])
          end
      end,
  })
````
EDIT: If it's not possible, I'll probably write some function which wrap the commands made available by mason.nvim

**Reply by williamboman** [reply]
Yeo, check `:h mason-lspconfig`!

**Reply by ghost** [reply]
> Yeo, check `:h mason-lspconfig`!

Sorry, I'm not sure that I understand. You mean I can also use `ensure_installed` from mason-lspconfig for stuff like `shfmt` and then use `mason-registry` to get all installed packages, filter by CAT and then get the commands of these and pass them onto null-ls ? 
Won't the generic handler try and start them, or is it already limited by the LSP category.

**Reply by williamboman** [reply]
The mason-lspconfig module is exclusively only for lspconfig-related functionality. So anything in mason-lspconfig is strictly limited to LSP servers. Also I’m not sure I’m following the relationship of ensure_installed and getting all installed packages? In the context of the null-ls question, let’s pretend mason-lspconfig doesnt exist. Yeah sure that sounds doable. I’m not sure you’ll need to specify any commands yourself as this is all handled by null-ls - I guess the only thing you need to do is map a Mason package to a null-ls ”source”? Something like:

```lua
— Maps a Mason package name to a null-ls source
local source_map = {
    stylua = require("null-ls").builtins.formatting.stylua,
    cspell = require("null-ls").builtins.diagnostics.cspell
}

local sources = {}

for _, pkg in ipairs(mason_registry.get_installed_package_names()) do
  if source_map[pkg_name] then
     table.insert(sources, source_map[pkg_name])
  end
end

require("null-ls").setup({
    sources = sources
})

```

On mobile atm so prob a few mistakes but hopefully you get the idea.

edit: this could probably turn into a very helpful mason-null-ls plugin 🤷‍♂️

**Reply by ghost** [reply]
>The mason-lspconfig module is exclusively only for lspconfig-related functionality. So anything in mason-lspconfig is strictly limited to LSP servers. 

That is intuitive and the obvious answer, but it wasn't clear 100% to me, thank you. I thought it might also be used to install other Mason packages, not just LSPs.
>Also I’m not sure I’m following the relationship of ensure_installed and getting all installed packages? In the context of the null-ls question, let’s pretend mason-lspconfig doesnt exist.

I'm looking for a way to not install them manually, as you can see from this piece of lua I posted above:
```lua
 local formatters = { "shfmt", "stylua" }

  for _, name in ipairs(formatters) do
      local formatter_is_found = format_install.exists(name)
      if formatter_is_found then
          if not format_install.is_installed(name) then
              format_install.install_formatter(name)
          end
      end
  end -- automatically install all formatters

  for _, formatter in ipairs(format_install.get_installed_formatters()) do
      if not util.set_contains_val(formatters, formatter.name) then
          format_install.uninstall_formatter(formatter.name)
      end
  end -- automatically uninstall all not needed formatters
```
My ideal case would be something like this that would also include a part for updating Mason packages. (I believe it was mentioned in another post of yours on the nvim-lsp-installer discussion / issue github section). 

To expand a little on this, I would like to just have a list somewhere in my config where I can declare what mason packages I want and then have a few lines of code to handle if they should be installed, if a mason package is present but should be uninstalled or to update them without having to call `:MasonInstall/MasonUninstall <package> ...` manually.

Then with all the Mason packages in order, if possible, I would create separate lists based on the category of the mason package and eventually make null-ls work nice with them using my initial approach or something similar to how you did it except maybe add a prefix because some software has multiple null-ls `builtins` (example: shellcheck which has both `linter` and `code_action`)

**Reply by nyngwang** [reply]
@williamboman I have a very long time not following this project. Is the code answer here stable now? If not where could I find the stable version?
